### ISSUE TITLE
Add Daylight-Savings Time (DST) algorithm

### ISSUE THREAD

**Issue by Emile666** [Emile666/Nixie_Board#2]
I have a working algorithm for DST. Add it to the code-base and adjust the DS3231 automatically for it.

**Comment by Emile666** [comment]
To start DST: Find the last Sunday in March: @2 AM advance clock to 3 AM.
To stop DST: Find the last Sunday in October: @3 AM set clock back to 2 AM (only once!).

**Comment by xlentelectronics** [comment]
I have a routine already available for DST.
I use it in my Wheater-Clock station and works fine! Only gap is when you set manualy the time on the last sunday in March or October between 02:00 and 03:00 hours. The system is than not able to detect if winter or summertime is already set.

This routine only works in the European time zone. The USA ueses another month and date for their DST.

This runs in the main program

```
  // Summertime or Wintertime - DST
  // Set Summertime
  if (WD == 1 && MM == 3 && DD >= 25 && DD <= 31 && hh == 2 && DST == false)
    {
      hh = 3;
      RTC.setHours(hh);
      RTC.setClock();
      DST = true;
    }      

  // Set Wintertime
  if (WD == 1 && MM == 10 && DD >= 25 && DD <= 31 && hh == 3 && DST == true)
    {
      hh = 2;
      RTC.setHours(hh);
      RTC.setClock();
      DST = false;
    }

This function is called when power is switch on and also when time & date are set.

// **********************************************
// *                                                             *
// *   Detect SummerTime or WinterTime *
// *                                                             *
//**********************************************

void DST_Detect()
{
  if (MM < 3 || MM > 10)      // It is Wintertime
  {
    DST = false;
  }
  else                        // It is Summertime if (MM > 3 && MM <10)
  {
    DST = true;
  } //End if  

  prevSunday = DD - WD;      // Date minus Weekday is prevSunday     

  if (MM == 3 && prevSunday < 25)          // Is the last Sunday in March already past? If so, than it is already Summertime   
  {                                        // Mismatch when system is powerd-up on the last Sunday of March. Summertime will not be set automatically. 
    DST = false;  // It is Wintertime
  }

  if (MM == 10 && prevSunday >= 24)        // Is the last Sunday in October already past? If so, than it is already Wintertime 
  {                                        // Mismatch when system is powerd-up on the last Sunday of October. Wintertime will not be set automatically.
    DST = false;  // It is Wintertime
  } //End if


  //Display for 3 seconds if it is SummerTime or WinterTime
  lcd.setCursor(0,0);
  lcd.print ("    It is....   ");
  lcd.setCursor(0,1);

  if (DST == false)
  {
    lcd.print ("   WinterTime!  ");
  }
  else
  {
    lcd.print ("   SummerTime!  ");
  } //End if      

  delay (3000);  

  lcd.clear();
  delay (100);
  lcd.home();

} //End DST_Detect
```

**Comment by Emile666** [comment]
Compare this to my version :) Note that the dow() function is already present in the code-base (is called ds3231_calc_dow()).

```
uint8_t dst_active(int y, int m, int d)
{
    int day,lsun03,lsun10;

    day    = dow(y,3,31);        // Find day-of-week for March 31th
    lsun03 = 31 - ((day+1) % 7); // Find last sunday in March
    day    = dow(y,10,31);       // Find day-of-week for October 31th
    lsun10 = 31 - ((day+1) % 7); // Find last sunday in October

    if (((m >   3) && (m < 10))      ||
        ((m ==  3) && (d >= lsun03)) ||
        ((m == 10) && (d <  lsun10)))
       return 1;
    return 0;
} // dst_active()

```

**Comment by Emile666** [comment]
Closed in commit 7c64fcc
